### PR TITLE
[WORKFLOWS-480] Include input/output names to the manifest format

### PR DESCRIPTION
**Pull request by BrunoGrandePhD**
This PR builds on top of #4 and improves the manifest format by restructuring the inputs and outputs of each task. Now, the names (including any `emit` names) are included in the output. Unfortunately, I'm not able to get these consistently, but I think this is a good place to start. I also "normalized" the output so that the tasks aren't duplicated under the published files. Instead, I'm using task IDs as "foreign keys". Lastly, I updated the test workflow to better feature the input/output name functionality. 

Here's an example of the latest format: [manifest.json](https://github.com/Sage-Bionetworks-Workflows/nf-prov/files/10045446/manifest.json.txt)

### CODE REVIEW

**Review by thomasyu888 — APPROVED** [review]
Awesome!  I was able to follow the instructions under the Developer section of the readme and re-create this.  Couple things:

* I had to change some of the code in `launch.sh` to be:  `../nextflow-nf-prov/launch.sh` using the built `nextflow-nf-prov` repo 
* I see what you mean about `emit`.  I'm curious as to why the tuple output isn't being emitted as an array with `emit: value`
* Great recursive use of `jsonify` and addition of `toUriString()`
* I'm not a groovy expert by any means, but I did multiple read throughs and it looks good. 
* Since you excluded the `$`, just out of curiosity, did they ever have values associated with them or was it just blank?



<details>
<summary>Here is the output</summary>

```
{
    "pipeline": {
        "name": "nf-prov-test",
        "author": "Bruno Grande",
        "homePage": "https://github.com/sage-bionetworks-workflows/nf-prov",
        "description": "Nextflow plugin for tracking provenance",
        "mainScript": "main.nf",
        "nextflowVersion": "!>=21.10.3",
        "version": "0.2.0",
        "doi": ""
    },
    "published": [
        {
            "source": "/Users/tyu/sage/nf-prov/work/d3/0e56ecdf5df817b7547c4800313d56/r3.foo.txt",
            "target": "/Users/tyu/sage/nf-prov/out/r3.foo.txt",
            "publishingTaskId": "3"
        },
        {
            "source": "/Users/tyu/sage/nf-prov/work/b5/26a4dce6b9ab1338b90704cd97286e/r1.foo.txt",
            "target": "/Users/tyu/sage/nf-prov/out/r1.foo.txt",
            "publishingTaskId": "1"
        },
        {
            "source": "/Users/tyu/sage/nf-prov/work/4d/7c91fa922f11f1c892629f5e98c14e/r2.foo.txt",
            "target": "/Users/tyu/sage/nf-prov/out/r2.foo.txt",
            "publishingTaskId": "2"
        }
    ],
    "tasks": {
        "3": {
            "id": "3",
            "name": "RNG (3)",
            "cached": false,
            "process": "RNG",
            "inputs": [
                {
                    "name": "prefix",
                    "value": "r3"
                },
                {
                    "name": "constant",
                    "value": "foo"
                }
            ],
            "outputs": [
                {
                    "name": "prefix",
                    "emit": null,
                    "value": "r3"
                },
                {
                    "name": "constant",
                    "emit": null,
                    "value": "foo"
                },
                {
                    "name": null,
                    "emit": "file",
                    "value": "/Users/tyu/sage/nf-prov/work/d3/0e56ecdf5df817b7547c4800313d56/r3.foo.txt"
                }
            ]
        },
        "1": {
            "id": "1",
            "name": "RNG (1)",
            "cached": false,
            "process": "RNG",
            "inputs": [
                {
                    "name": "prefix",
                    "value": "r1"
                },
                {
                    "name": "constant",
                    "value": "foo"
                }
            ],
            "outputs": [
                {
                    "name": "prefix",
                    "emit": null,
                    "value": "r1"
                },
                {
                    "name": "constant",
                    "emit": null,
                    "value": "foo"
                },
                {
                    "name": null,
                    "emit": "file",
                    "value": "/Users/tyu/sage/nf-prov/work/b5/26a4dce6b9ab1338b90704cd97286e/r1.foo.txt"
                }
            ]
        },
        "2": {
            "id": "2",
            "name": "RNG (2)",
            "cached": false,
            "process": "RNG",
            "inputs": [
                {
                    "name": "prefix",
                    "value": "r2"
                },
                {
                    "name": "constant",
                    "value": "foo"
                }
            ],
            "outputs": [
                {
                    "name": "prefix",
                    "emit": null,
                    "value": "r2"
                },
                {
                    "name": "constant",
                    "emit": null,
                    "value": "foo"
                },
                {
                    "name": null,
                    "emit": "file",
                    "value": "/Users/tyu/sage/nf-prov/work/4d/7c91fa922f11f1c892629f5e98c14e/r2.foo.txt"
                }
            ]
        }
    }
}
```

</details>

File: plugins/nf-prov/src/main/nextflow/prov/ProvObserver.groovy [hunk 1]
```diff
@@ -151,21 +169,27 @@ class ProvObserver implements TraceObserver {
...
         // generate temporary output-task map
         def outputTaskMap = [:]
-        this.tasks.each { task ->
-            task.outputs.each { output ->
-                outputTaskMap.put(output, task)
+        this.tasks.each { taskId, task ->
+            task['outputs'].each { output ->
+                // Make sure to handle tuples of outputs
+                def values = output['value']
+                if ( values instanceof Collection ) {
```
[the remark this answers is not in the document]
Comment (BrunoGrandePhD) [reply plugins/nf-prov/src/main/nextflow/prov/ProvObserver.groovy:182]: What change did you make to `test.nf`?

I changed the `RNG` process to this:

```nextflow
process RNG {

    publishDir "out/", mode: 'copy'

    input:
    tuple val(prefix), val(constant)

    output:
    tuple val(prefix), val(constant), emit: 'values'
    tuple path("*.txt"), path("*.csv"), emit: 'file'

    script:
    """
    echo \$RANDOM > ${prefix}.${constant}.txt
    echo \$RANDOM > ${prefix}.${constant}.csv
    """
}
```

And I got this as my published files:

```json
    "published": [
        {
            "source": "/Users/bgrande/Repos/nf-prov-project/nf-prov/work/a8/ad44e4e46fffa0d1733e2a7ff791ce/r2.foo.csv",
            "target": "/Users/bgrande/Repos/nf-prov-project/nf-prov/out/r2.foo.csv",
            "publishingTaskId": "2"
        },
        {
            "source": "/Users/bgrande/Repos/nf-prov-project/nf-prov/work/a8/ad44e4e46fffa0d1733e2a7ff791ce/r2.foo.txt",
            "target": "/Users/bgrande/Repos/nf-prov-project/nf-prov/out/r2.foo.txt",
            "publishingTaskId": "2"
        },
        {
            "source": "/Users/bgrande/Repos/nf-prov-project/nf-prov/work/34/ed3edf6cbefa938c856c281d21e96b/r3.foo.csv",
            "target": "/Users/bgrande/Repos/nf-prov-project/nf-prov/out/r3.foo.csv",
            "publishingTaskId": "3"
        },
        {
            "source": "/Users/bgrande/Repos/nf-prov-project/nf-prov/work/34/ed3edf6cbefa938c856c281d21e96b/r3.foo.txt",
            "target": "/Users/bgrande/Repos/nf-prov-project/nf-prov/out/r3.foo.txt",
            "publishingTaskId": "3"
        },
        {
            "source": "/Users/bgrande/Repos/nf-prov-project/nf-prov/work/ef/a3afb6132938b750b21f3fddb0e378/r1.foo.csv",
            "target": "/Users/bgrande/Repos/nf-prov-project/nf-prov/out/r1.foo.csv",
            "publishingTaskId": "1"
        },
        {
            "source": "/Users/bgrande/Repos/nf-prov-project/nf-prov/work/ef/a3afb6132938b750b21f3fddb0e378/r1.foo.txt",
            "target": "/Users/bgrande/Repos/nf-prov-project/nf-prov/out/r1.foo.txt",
            "publishingTaskId": "1"
        }
    ]
```

**Comment by BrunoGrandePhD** [comment]
I'm glad the README instructions worked for you. I updated the `launch.sh` file to work with them. 

> I'm curious as to why the tuple output isn't being emitted as an array with emit: value

I suspect that it has to do with the way that Nextflow handles inputs/outputs in a task run vs the process definition. I might ping Paolo for help with this one once I've done a bit more research into the various Groovy classes. 

> Since you excluded the $, just out of curiosity, did they ever have values associated with them or was it just blank?

It was always `true`. I need to figure out what it means, but I'm omitting it for now.

File: plugins/nf-prov/src/main/nextflow/prov/ProvObserver.groovy (hunk 1 above)
[the remark this answers is not in the document]
Comment (thomasyu888) [reply plugins/nf-prov/src/main/nextflow/prov/ProvObserver.groovy:182]: The second echo was
```
    echo \$RANDOM > random.${prefix}.${constant}.csv
```

or you end up writing to the same file.

[the remark this answers is not in the document]
Comment (BrunoGrandePhD) [reply plugins/nf-prov/src/main/nextflow/prov/ProvObserver.groovy:182]: I had changed the file extension from txt to csv. How did you structure your output line?

[the remark this answers is not in the document]
Comment (thomasyu888) [reply plugins/nf-prov/src/main/nextflow/prov/ProvObserver.groovy:182]: ahh, I didn't even notice that!  sorry about that...
```
process RNG {

    publishDir "out/", mode: 'copy'

    input:
    tuple val(prefix), val(constant)

    output:
    tuple val(prefix), val(constant), emit: 'values'
    path "*.txt", emit: 'file'

    script:
    """
    echo \$RANDOM > ${prefix}.${constant}.txt
    echo \$RANDOM > doo.${prefix}.${constant}.txt

    """

}
```

[the remark this answers is not in the document]
Comment (BrunoGrandePhD) [reply plugins/nf-prov/src/main/nextflow/prov/ProvObserver.groovy:182]: Ah, got it! I've permanently modified `test.nf` to include this "edge case".
